File: java/Mandelbrot.java
class Mandelbrot {
    private static final char[] chars = "$@B%8&WM#*|1?-_+~i!I:,'. ".toCharArray();

    public static void main(String[] args) {
        for (double a = -2; a <= 2; a += 0.05) {
            for (double bi = -2; bi <= 2; bi += 0.05) {
                Complex c = new Complex(a, bi);
                int m = mandelbrot(new Complex(0, 0), c, 0, 1000);
                System.out.print(getChar(m));
            }
            System.out.println();
        }
    }

    private static char getChar(int i) {
        return chars[i % 25];
    }

    private static int mandelbrot(Complex z, Complex c, int n, int m) {
        if (Complex.magnitude(z) >= 2 || n == m) {
            return n;
        } else {
            return mandelbrot(
                Complex.add(Complex.multiply(z, z), c), 
                c,
                n + 1, 
                m);
        }
    }
}

File: haskell/mandelbrot.hs
import Data.Complex
import Data.List

main = putStrLn $ intercalate "\n" grid

mandelbrot z m n c 
    | magnitude z >= 2 || n == m = n
    | otherwise = mandelbrot (z * z + c) m (n + 1) c

grid = [[toChar $ escapeTime (a :+ 2 * bi) | a <- bounds_a] | bi <- bounds_bi]
    where
        escapeTime = mandelbrot 0 1000 0
        toChar c   = "$@B%8&WM#*|1?-_+~i!I:,'. " !! (mod c 25)
        bounds_a   = [-2,-1.95..2]
        bounds_bi  = [-2,-1.95..2]
